cdap-app-fabric/src/test/java/io/cdap/cdap/master/environment/DefaultMasterEnvironmentRunnableContextTest.java

```diff
@@ -0,0 +1,118 @@
+import com.google.inject.Injector;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.discovery.ResolvingDiscoverable;
+import io.cdap.cdap.common.discovery.URIScheme;
+import io.cdap.cdap.common.guice.LocalLocationModule;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.HttpResponder;
+import io.cdap.http.NettyHttpService;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.discovery.DiscoveryService;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.discovery.InMemoryDiscoveryService;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.nio.charset.StandardCharsets;
+import java.util.stream.Collectors;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+/**
+ * Test for {@link DefaultMasterEnvironmentRunnableContextTest}.
+ */
+public class DefaultMasterEnvironmentRunnableContextTest {
+  private static DefaultMasterEnvironmentRunnableContext context;
+  private static DiscoveryService discoveryService;
+  private static Cancellable cancelDiscovery;
+  private static NettyHttpService httpService;
+
+  @BeforeClass
+  public static void setup() throws Exception {
+    discoveryService = new InMemoryDiscoveryService();
+    Injector injector = Guice.createInjector(
+      new LocalLocationModule(),
+      new AbstractModule() {
+        @Override
+        protected void configure() {
+          bind(CConfiguration.class).toInstance(CConfiguration.create());
+          bind(DiscoveryServiceClient.class).toInstance((DiscoveryServiceClient) discoveryService);
+        }
+      });
+    context = injector.getInstance(DefaultMasterEnvironmentRunnableContext.class);
+
+    httpService = NettyHttpService.builder(Constants.Service.APP_FABRIC_HTTP)
+      .setHttpHandlers(new MockHttpHandler())
+      .build();
+    httpService.start();
+    cancelDiscovery = discoveryService.register(
+      ResolvingDiscoverable.of(URIScheme.createDiscoverable(Constants.Service.APP_FABRIC_HTTP, httpService)));
+  }
+
+  @AfterClass
+  public static void stop() throws Exception {
+    if (cancelDiscovery != null) {
+      cancelDiscovery.cancel();
+    }
+    if (httpService != null) {
+      httpService.stop();
+    }
+  }
+
+  @Test
+  public void testOpenHttpConnection() throws Exception {
+    String message = "hello";
+    String resource = String.format("echo/%s", message);
+    HttpURLConnection conn = context.openHttpURLConnection(resource);
+    conn.setRequestMethod(HttpMethod.GET.name());
+    InputStream is = conn.getInputStream();
```

Comment:
This is leaking resource. Better just use `HttpRequests` class to help writing the test?

Reply:
Wrapped in a try with resource. 

For using HttpRequests, none of its API takes a connection. Creating, opening and using connection are all private and within the HttpRequests.